.NET/Microsoft.Recognizers.Text.DateTime/Extractors/BaseDatePeriodExtractor.cs

```diff
@@ -57,17 +83,91 @@ private List<Token> MatchSimpleCases(string text)
             return ret;
         }
 
+        // Complex cases refer to the combination of daterange and datepoint
+        // For Example: from|between {DateRange|DatePoint} to|till|and {DateRange|DatePoint}
+        private List<Token> MatchComplexCases(string text, List<ExtractResult> simpleDateRangeResults, DateObject reference)
+        {
+            var ret = new List<Token>();
+            var er = this.config.DatePointExtractor.Extract(text, reference);
+
+            // Filter out DateRange results that are part of DatePoint results
+            // For example, "Feb 1st 2018" => "Feb" and "2018" should be filtered out here
+            er.AddRange(simpleDateRangeResults
+                .Where(simpleDateRange => !er.Any(datePoint => (datePoint.Start <= simpleDateRange.Start && datePoint.Start + datePoint.Length >= simpleDateRange.Start + simpleDateRange.Length))));
+
+            er = er.OrderBy(t => t.Start).ToList();
+
+            if (er.Count <= 1)
+            {
+                return ret;
+            }
+
+            var idx = 0;
+
+            while (idx < er.Count - 1)
+            {
+                var middleBegin = er[idx].Start + er[idx].Length ?? 0;
+                var middleEnd = er[idx + 1].Start ?? 0;
+                if (middleBegin >= middleEnd)
+                {
+                    idx++;
+                    continue;
+                }
+
+                var middleStr = text.Substring(middleBegin, middleEnd - middleBegin).Trim().ToLowerInvariant();
+                var match = this.config.TillRegex.Match(middleStr);
+                if (match.Success && match.Index == 0 && match.Length == middleStr.Length)
+                {
+                    var periodBegin = er[idx].Start ?? 0;
+                    var periodEnd = (er[idx + 1].Start ?? 0) + (er[idx + 1].Length ?? 0);
+
+                    // handle "desde"
+                    var beforeStr = text.Substring(0, periodBegin).Trim().ToLowerInvariant();
+                    if (this.config.GetFromTokenIndex(beforeStr, out int fromIndex)
+                        || this.config.GetBetweenTokenIndex(beforeStr, out fromIndex))
+                    {
+                        periodBegin = fromIndex;
+                    }
+
+                    ret.Add(new Token(periodBegin, periodEnd));
+                    idx += 2;
```

Comment:
Can you add a comment everywhere "magic" numbers like this happen? The comment should describe why 2.

Reply:
Added comments to describe this "magic" number.